MakieCore/src/recipes.jl

```diff
@@ -189,6 +189,381 @@ macro recipe(theme_func, Tsym::Symbol, args::Symbol...)
+        # recipe, so that we can offer the convenience of automatic augmented docstrings
+        # but combine them with the simplicity of using a normal docstring.
+        # The trick is to mark some variable (in this case a gensymmed placeholder) with the
+        # Core.@__doc__ macro, which causes this variable to get assigned the docstring on top
+        # of the @recipe invocation. From there, it can then be retrieved, modified, and later
+        # attached to plotting function by using @doc again. We also delete the binding to the
+        # temporary variable so no unnecessary docstrings stay in place.
+        Core.@__doc__ $(esc(docs_placeholder)) = nothing
+        binding = Docs.Binding(@__MODULE__, $(QuoteNode(docs_placeholder)))
+        user_docstring = if haskey(Docs.meta(@__MODULE__), binding)
+            _docstring = @doc($docs_placeholder)
+            delete!(Docs.meta(@__MODULE__), binding)
+            _docstring
+        else
+            "No docstring defined.\n"
+        end
+
+
+        $(funcname)() = not_implemented_for($funcname)
+        const $(PlotType){$(esc(:ArgType))} = Plot{$funcname,$(esc(:ArgType))}
+
+        # This weird syntax is so that the output of the macrocall can be escaped because it
+        # contains user expressions, without escaping what's passed to the macro because that
+        # messes with its transformation logic. Because we escape the whole block with the macro,
+        # we don't reference it by symbol but splice in the macro itself into the AST
+        # with `var"@DocumentedAttributes"`
+        const $attr_placeholder = $(
+            esc(Expr(:macrocall, var"@DocumentedAttributes", LineNumberNode(@__LINE__), attrblock))
+        )
+
+        $(MakieCore).documented_attributes(::Type{<:$(PlotType)}) = $attr_placeholder
+
+        $(MakieCore).plotsym(::Type{<:$(PlotType)}) = $(QuoteNode(Tsym))
+        function ($funcname)(args...; kw...)
+            kwdict = Dict{Symbol, Any}(kw)
+            _create_plot($funcname, kwdict, args...)
+        end
+        function ($funcname!)(args...; kw...)
+            kwdict = Dict{Symbol, Any}(kw)
+             _create_plot!($funcname, kwdict, args...)
+        end
+        function MakieCore.is_attribute(T::Type{<:$(PlotType)}, sym::Symbol)
+            sym in MakieCore.attribute_names(T)
+        end
+
+        function MakieCore.attribute_names(T::Type{<:$(PlotType)})
+            keys(documented_attributes(T).d)
+        end
+
+        function $(MakieCore).default_theme(scene, T::Type{<:$PlotType})
+            Attributes(documented_attributes(T).closure(scene))
+        end
+
+        function MakieCore.attribute_default_expressions(T::Type{<:$PlotType})
+            Dict(k => v.default_expr for (k, v) in documented_attributes(T).d)
+        end
+
+        function MakieCore._attribute_docs(T::Type{<:$PlotType})
+            Dict(k => v.docstring for (k, v) in documented_attributes(T).d)
+        end
```

Comment:
```julia
        function MakieCore.is_attribute(T::Type{<:$(PlotType)}, sym::Symbol)
            sym in MakieCore.attribute_names(T)
        end

        function MakieCore.attribute_names(T::Type{<:$(PlotType)})
            keys(documented_attributes(T).d)
        end

        function $(MakieCore).default_theme(scene, T::Type{<:$PlotType})
            Attributes(documented_attributes(T).closure(scene))
        end

        function MakieCore.attribute_default_expressions(T::Type{<:$PlotType})
            Dict(k => v.default_expr for (k, v) in documented_attributes(T).d)
        end

        function MakieCore._attribute_docs(T::Type{<:$PlotType})
            Dict(k => v.docstring for (k, v) in documented_attributes(T).d)
        end
```
These functions could be defined outside the macro, can they?

Reply:
true, good catch